### PR TITLE
Interface based API

### PR DESCRIPTION
**Pull request by rorbech**
DO NOT PUSH DIRECTLY TO THIS PR. All changes should be through other PRs.

This PR is an umbrella for making the API purely interface based

TODO 
- [x] Split public classes into interface and implementation 
  - [x] All `Realm` related - #376 
  - [x] List APIs - #396
- [ ] Evaluate if we should also split data classes (currently LogConfiguration and VersionID)
- [x] Commonize `version` and various lifecycle implementations - Separate PR in #383  
- [ ] Internalize non-public API - Should be in a separate PR #56 
- [ ] Enable explicit API mode - Should be in a separate PR #22
- [ ] Consider interface options for RealmObject (**Postponed for later** as we don't have the ideal fix for this yet - already tracked by various issues i.e. #83, #203 and #219)
       - We cannot just add interface methods to `RealmObject` and implement them internally as IDE will consider these unimplemented due to the invisible `RealmObject->RealmObjectInternal` relationship injected by the compiler plugin.
       - Adding methods with default implementation is an option, but cannot directly dispatch to internal `RealmObjectInternal` implementation as it generates recursive calls. Could just dispatch to the `realmState().version()` instead of `RealmObjectInternal`.
- [x] Update CHANGLOG, README and other docs

### CODE REVIEW

File: packages/plugin-compiler/src/main/kotlin/io/realm/compiler/RealmSchemaLoweringExtension.kt [hunk 1]
```diff
@@ -70,73 +69,73 @@ import org.jetbrains.kotlin.name.Name
...
-                            putValueArgument(2, populatedCompanionMap)
-                        }
-                    }
-                    return super.visitConstructorCall(expression)
-                }
-
                 override fun visitCall(expression: IrCall): IrExpression {
-                    if (expression.symbol.owner.name == Name.identifier("build") &&
+                    val name = expression.symbol.owner.name
+                    if (name in setOf(REALM_CONFIGURATION_BUILDER_BUILD, REALM_CONFIGURATION_DEFAULT_CONFIG) &&
```
Comment (nhachicha) [inline packages/plugin-compiler/src/main/kotlin/io/realm/compiler/RealmSchemaLoweringExtension.kt:85]: Nitpick: I think checking the FQNAME first is more optimal
```Kotlin
expression.type.classFqName == REALM_CONFIGURATION && name in setOf(REALM_CONFIGURATION_BUILDER_BUILD, REALM_CONFIGURATION_DEFAULT_CONFIG
```

File: packages/plugin-compiler/src/main/kotlin/io/realm/compiler/RealmSchemaLoweringExtension.kt [hunk 2]
```diff
@@ -70,73 +69,73 @@ import org.jetbrains.kotlin.name.Name
...
-                            // substitute build function with internal one taking the companion map as argument
-                            val internalBuildFunction = pluginContext.lookupClassOrThrow(REALM_CONFIGURATION_BUILDER).functions.first {
-                                it.name == Name.identifier("build") && it.valueParameters.size == 1
+                        val specifiedModels =
+                            mutableListOf<Triple<IrClassifierSymbol, IrType, IrClassSymbol>>()
+                        val (receiver, schemaArgument) = when (name) {
+                            // Replaces `RealmConfiguration.Builder.build()` with RealmConfiguration.Builder(...).build(companionMap)
+                            REALM_CONFIGURATION_BUILDER_BUILD -> {
+                                val builder = expression.dispatchReceiver
+                                val schemaArgument = expression
```
Comment (nhachicha) [inline packages/plugin-compiler/src/main/kotlin/io/realm/compiler/RealmSchemaLoweringExtension.kt:94]:
```suggestion
```

File: packages/plugin-compiler/src/main/kotlin/io/realm/compiler/RealmSchemaLoweringExtension.kt [hunk 3]
```diff
@@ -70,73 +69,73 @@ import org.jetbrains.kotlin.name.Name
...
+                                builder to schemaArgument
```
Comment (nhachicha) [inline packages/plugin-compiler/src/main/kotlin/io/realm/compiler/RealmSchemaLoweringExtension.kt:95]: inline variable
```suggestion
                                builder to expression
```

**Review by nhachicha — COMMENTED** [review]
Nice refactor 👍  a couple of minors comments

File: packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt [hunk 4]
```diff
@@ -128,102 +110,26 @@ public class RealmConfiguration private constructor(
...
+    companion object {
+        /**
+         * Create a configuration using default values except for schema, path and name.
+         *
+         * @param path The full path of the realm file.
+         * @param name The filename of the realm file.
+         * @param schema The classes of the schema. The elements of the set must be direct class literals.
+         */
+        // Should always follow Builder constructor arguments
+        fun defaultConfig(
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt:124]: @cmelchior Any hard feelings about this?

[the remark this answers is not in the document]
Comment (cmelchior) [reply packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt:124]: I feel pretty strongly that we need a helper better than `RealmConfiguration.Builder(schema).build()`, but I have less strong opinions about what it is called. I do follow Nabils reasoning, but I think the `default` was mostly to infer that the rest of the config was default.

After playing around with the Sync API's, we are going to have quite a lot of API factories

```
SyncConfiguration.defaultConfig(user, partition, schema)
RealmConfiguration.defaultConfig(schema)

Realm.open(config)

AppConfiguration.create(appId)

App.create(config)
App.create(appId)

realmListOf()
```

So if it was possible to use the same pattern for all of them, that would be nice, but maybe that is stretching it

But to the suggestion:
```
AppConfiguration.with("appId")
SyncConfiguration.with(user, partitionValue, schema)
RealmConfiguration.with(user, partitionValue, schema)
```

Doesn't feel that far-fetched, no, and does sort of imply the same `with` behaviour as from the the Kotlin stdlib.

[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library/src/commonMain/kotlin/io/realm/RealmConfiguration.kt]: Rename from `defaultConfig` to `with`.

File: packages/plugin-compiler/src/main/kotlin/io/realm/compiler/RealmModelLoweringExtension.kt [hunk 5]
```diff
@@ -42,8 +51,10 @@ private class RealmModelLowering(private val pluginContext: IrPluginContext) : C
     override fun lower(irClass: IrClass) {
         if (irClass.hasRealmModelInterface) {
             // add super type RealmObjectInternal and RealmObjectInterop
-            val realmObjectInteropInterface: IrClassSymbol = pluginContext.lookupClassOrThrow(REALM_OBJECT_INTEROP_INTERFACE).symbol
-            val realmObjectInternalInterface: IrClassSymbol = pluginContext.lookupClassOrThrow(REALM_OBJECT_INTERNAL_INTERFACE).symbol
+            val realmObjectInteropInterface: IrClassSymbol =
+                pluginContext.lookupClassOrThrow(REALM_OBJECT_INTEROP_INTERFACE).symbol
+            val realmObjectInternalInterface: IrClassSymbol =
+                pluginContext.lookupClassOrThrow(REALM_OBJECT_INTERNAL_INTERFACE).symbol
             irClass.superTypes += realmObjectInteropInterface.defaultType
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/plugin-compiler/src/main/kotlin/io/realm/compiler/RealmModelLoweringExtension.kt:58]: Targeted by #442

**Comment by rorbech** [comment]
I have targeted the comments in #442

**Review by nhachicha — APPROVED** [review]

File: packages/library/src/commonMain/kotlin/io/realm/internal/RealmLifeCycle.kt [hunk 6]
```diff
@@ -17,12 +17,44 @@
...
  */
-internal interface RealmLifeCycle {
-    fun version(): VersionId
+internal interface RealmLifeCycle : Versioned {
     fun isFrozen(): Boolean
     fun isClosed(): Boolean
 }
+
+// Singleton instance acting as implementation for all unmanaged objects
+object UnmanagedLifeCycle : RealmLifeCycle {
```
[the remark this answers is not in the document]
Comment (rorbech) [reply packages/library/src/commonMain/kotlin/io/realm/internal/RealmLifeCycle.kt]: Renamed to `RealmState` in #442.

**Pull request merged**